File: 2019/6/solution.cpp
#include <vector>
#include <string>
#include <sstream>
#include <iostream>
#include <fstream>
#include <unordered_map>
#include <numeric>
#include <unordered_set>
#include <deque>
#include <algorithm>

using namespace std;

struct Edge {
  string a;
  string b;
};


int main() {
  vector<Edge> edges;
  unordered_set<string> nodes;
  fstream stream ("input");
  string line;
  while (getline(stream, line)) {
    auto idx = line.find(')');
    auto a = line.substr(0, idx);
    auto b = line.substr(idx + 1);
    edges.emplace_back(Edge{a, b});
    nodes.emplace(a);
    nodes.emplace(b);
  }

  unordered_map<string, int> distance;
  for (const auto& n : nodes) {
    distance.try_emplace(n, numeric_limits<int>::max());
  }

  unordered_set<string> visited;
  deque<string> q;
  distance["YOU"] = 0;
  q.emplace_back("YOU");
  while (!q.empty()) {
    auto n = q.front();
    q.pop_front();
    visited.emplace(n);
    for (const auto& [a, b] : edges) {
      if ((a == n) || (b == n)) {
        auto other = (a == n) ? b : a;
        auto iter = visited.find(other);
        if (iter == visited.end()) {
          distance[other] = min(distance[other], distance[n] + 1);
          q.emplace_back(other);
        }
      }
    }
  }

  cout << distance["SAN"] - 2 << '\n';
}

File: 2019/6/solution.py
import math
from collections import deque
lines = [line.strip() for line in open('input').readlines()]

nodes = set()
edges = []
for line in lines:
    a, b = line.split(')')
    nodes.add(a)
    nodes.add(b)
    edges.append((a, b))


# def visit(node):
#     q = deque()
#     visited = set()
#     q.append(node)
#     while q:
#         n = q.popleft()
#         visited.add(n)
#         for a, b in edges:
#             if b == n and a not in visited:
#                 q.append(a)
#     return len(visited) - 1


# print(visit('C'))
# s = 0
# for node in nodes:
#     v = visit(node)
#     # print(node, v)
#     s += v
# print(s)

# print(visit())


distance = {}
for node in nodes:
    distance[node] = math.inf

distance['YOU'] = 0
visited = set()
q = deque()
q.append('YOU')
while q:
    n = q.popleft()
    visited.add(n)
    for a, b in edges:
        if a == n or b == n:
            other = a if b == n else b
            if other not in visited:
                distance[other] = min(distance[other], distance[n] + 1)
                q.append(other)


print(distance['SAN'] - 2)
# print(len(distance))
# l = [k for k, v in distance.items() if v != math.inf]
# print(len(l))

# for k, v in distance.items():
#     print(k, v)
